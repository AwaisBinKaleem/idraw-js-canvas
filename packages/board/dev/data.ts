import type { Data } from '@idraw/types';
import { createUUID, parseSVGPath } from '@idraw/util';

const data: Data = {
  elements: [
    // {
    //   uuid: createUUID(),
    //   type: 'image',
    //   x: 100,
    //   y: 100,
    //   w: 100,
    //   h: 100,
    //   angle: 30,
    //   detail: {
    //     // src: '/public/images/lena.png?t=002'
    //     src: '@assets/0ee02fb0-ec43-6b9a-9ab4-55be5816ca4a'
    //   },
    //   operations: {
    //     limitRatio: true
    //   }
    // },
    {
      uuid: createUUID(),
      x: 2,
      y: 2,
      w: 100,
      h: 100,
      type: 'circle',
      detail: {
        background: '#f44336'
      }
    },
    {
      uuid: createUUID(),
      type: 'rect',
      x: 50,
      y: 50,
      w: 100,
      h: 100,
      detail: {
        background: '#8bc34a'
      }
    },
    // {
    //   uuid: createUUID(),
    //   type: 'image',
    //   x: 250,
    //   y: 250,
    //   w: 100,
    //   h: 100,
    //   detail: {
    //     // src: '/public/images/github.png?t=003'
    //     src: '@assets/25a0e630-a958-a522-29a4-5dccf04985da'
    //   }
    // },
    // {
    //   uuid: createUUID(),
    //   type: 'group',
    //   x: 250,
    //   y: 250,
    //   w: 100,
    //   h: 100,
    //   detail: {
    //     children: [
    //       {
    //         uuid: createUUID(),
    //         type: 'image',
    //         x: 0,
    //         y: 0,
    //         w: 100,
    //         h: 100,
    //         detail: {
    //           // src: '/public/images/github.png?t=003'
    //           src: '@assets/489c64d2-f642-3855-7bd6-cb40624b62c3'
    //         }
    //       }
    //     ],
    //     // assets: {
    //     //   '@assets/489c64d2-f642-3855-7bd6-cb40624b62c3': {
    //     //     type: 'image',
    //     //     value: '/public/images/github.png?t=003'
    //     //   }
    //     // }
    //   }
    // },
    {
      uuid: createUUID(),
      x: 0,
      y: 300,
      w: 100,
      h: 100,
      type: 'circle',
      detail: {
        background: '#009688'
      }
    },
    {
      uuid: createUUID(),
      x: 300,
      y: 300,
      w: 100,
      h: 100,
      type: 'circle',
      detail: {
        background: '#673ab7'
      }
    },
    {
      uuid: createUUID(),
      x: 300,
      y: 0,
      w: 100,
      h: 100,
      type: 'circle',
      detail: {
        background: '#ffc107'
      }
    },
    {
      uuid: createUUID(),
      x: 150,
      y: 150,
      w: 100,
      h: 100,
      // angle: 30,
      type: 'rect',
      detail: {
        background: '#4caf50',
        clipPath: {
          commands: parseSVGPath(`
          M 10,30
          A 20,20 0,0,1 50,30
          A 20,20 0,0,1 90,30
          Q 90,60 50,90
          Q 10,60 10,30 z`),
          // fill: '#ff0000',
          originX: 10,
          originY: 10,
          originH: 80,
          originW: 80
          // background: '#0000002A'
        }
      }
    },
    {
      uuid: createUUID(),
      x: 100,
      y: 10,
      w: 80,
      h: 80,
      // angle: 30,
      type: 'path',
      detail: {
        commands: parseSVGPath(`
        M 10,30
        A 20,20 0,0,1 50,30
        A 20,20 0,0,1 90,30
        Q 90,60 50,90
        Q 10,60 10,30 z`),
        fill: 'red',
        originX: 10,
        originY: 10,
        originH: 80,
        originW: 80,
        background: '#0000002A'

        // clipPath: {
        //   commands: parseSVGPath(`
        //     M 10,30
        //     A 20,20 0,0,1 50,30
        //     A 20,20 0,0,1 90,30
        //     Q 90,60 50,90
        //     Q 10,60 10,30 z`),
        //   fill: '#ff0000',
        //   originX: 10,
        //   originY: 10,
        //   originH: 80,
        //   originW: 80
        // }
      }
    }
  ]
  // assets: {
  //   '@assets/0ee02fb0-ec43-6b9a-9ab4-55be5816ca4a': {
  //     type: 'image',
  //     value: '/public/images/lena.png?t=003'
  //   },
  //   '@assets/25a0e630-a958-a522-29a4-5dccf04985da': {
  //     type: 'image',
  //     value: '/public/images/github.png?t=002'
  //   }
  // }
};

export default data;
