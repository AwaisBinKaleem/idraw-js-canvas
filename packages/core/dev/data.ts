import type { Data } from '@idraw/types';
import { createUUID, parseSVGPath } from '@idraw/util';

const data: Data = {
  elements: [
    {
      uuid: createUUID(),
      type: 'image',
      x: 100,
      y: 100,
      w: 100,
      h: 100,
      angle: 30,
      detail: {
        src: '/images/lena.png?t=002'
      },
      operations: {
        limitRatio: true
      }
    },
    {
      uuid: createUUID(),
      x: 2,
      y: 2,
      w: 100,
      h: 100,
      type: 'circle',
      detail: {
        background: '#f44336'
      }
    },
    {
      uuid: createUUID(),
      type: 'rect',
      x: 50,
      y: 50,
      w: 100,
      h: 100,
      detail: {
        background: '#8bc34a'
      }
    },
    {
      uuid: createUUID(),
      type: 'image',
      x: 250,
      y: 250,
      w: 100,
      h: 100,
      detail: {
        src: '/images/github.png?t=003'
      }
    },
    {
      uuid: createUUID(),
      x: 0,
      y: 300,
      w: 100,
      h: 100,
      type: 'circle',
      detail: {
        background: '#009688'
      }
    },
    {
      uuid: createUUID(),
      x: 300,
      y: 300,
      w: 100,
      h: 100,
      type: 'circle',
      detail: {
        background: '#673ab7'
      }
    },
    {
      uuid: createUUID(),
      x: 300,
      y: 0,
      w: 100,
      h: 100,
      type: 'circle',
      detail: {
        background: '#ffc107'
      }
    },
    {
      uuid: createUUID(),
      x: 150,
      y: 150,
      w: 100,
      h: 100,
      angle: 30,
      type: 'rect',
      detail: {
        background: '#4caf50',
        clipPath: {
          commands: parseSVGPath(`
          M 10,30
          A 20,20 0,0,1 50,30
          A 20,20 0,0,1 90,30
          Q 90,60 50,90
          Q 10,60 10,30 z`),
          // fill: '#ff0000',
          originX: 10,
          originY: 10,
          originH: 80,
          originW: 80
          // background: '#0000002A'
        }
      }
    },
    {
      uuid: createUUID(),
      x: 100,
      y: 10,
      w: 80,
      h: 80,
      // angle: 30,
      type: 'path',
      detail: {
        commands: parseSVGPath(`
        M 10,30
        A 20,20 0,0,1 50,30
        A 20,20 0,0,1 90,30
        Q 90,60 50,90
        Q 10,60 10,30 z`),
        fill: 'red',
        originX: 10,
        originY: 10,
        originH: 80,
        originW: 80,
        background: '#0000002A'

        // clipPath: {
        //   commands: parseSVGPath(`
        //     M 10,30
        //     A 20,20 0,0,1 50,30
        //     A 20,20 0,0,1 90,30
        //     Q 90,60 50,90
        //     Q 10,60 10,30 z`),
        //   fill: '#ff0000',
        //   originX: 10,
        //   originY: 10,
        //   originH: 80,
        //   originW: 80
        // }
      }
    }
  ]
};

export default data;
