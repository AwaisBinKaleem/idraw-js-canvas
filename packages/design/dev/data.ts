import type { DesignData } from '../src';

const data: DesignData = {
  components: [
    {
      uuid: 'demo-xxx-001',
      type: 'component',
      name: 'demo',
      x: 50,
      y: 50,
      w: 100,
      h: 100,
      desc: {
        bgColor: '#1f1f1f',
        children: [
          {
            uuid: 'group-001-0014',
            type: 'circle',
            x: -40,
            y: 0,
            w: 100,
            h: 100,
            desc: {
              bgColor: '#f44336'
            }
          },
          {
            uuid: 'group-001-0015',
            type: 'circle',
            x: -20,
            y: 0,
            w: 100,
            h: 100,
            desc: {
              bgColor: '#ff9800'
            }
          },
          {
            uuid: 'group-001-0016',
            type: 'circle',
            x: 0,
            y: 0,
            w: 100,
            h: 100,
            desc: {
              bgColor: '#ffc106'
            }
          },
          {
            uuid: 'group-001-0017',
            type: 'circle',
            x: 20,
            y: 0,
            w: 100,
            h: 100,
            desc: {
              bgColor: '#cddc39'
            }
          },
          {
            uuid: 'group-001-0018',
            type: 'circle',
            x: 40,
            y: 0,
            w: 100,
            h: 100,
            desc: {
              bgColor: '#4caf50'
            }
          }
        ]
      }
    },
    {
      uuid: 'demo-xxx-002',
      type: 'component',
      name: 'demo',
      x: 50,
      y: 50,
      w: 100,
      h: 100,
      desc: {
        bgColor: '#f0f0f0',
        children: [
          {
            uuid: 'group-001-0014',
            type: 'circle',
            x: -40,
            y: 0,
            w: 100,
            h: 100,
            desc: {
              bgColor: '#f44336'
            }
          },
          {
            uuid: 'group-001-0015',
            type: 'circle',
            x: -20,
            y: 0,
            w: 100,
            h: 100,
            desc: {
              bgColor: '#ff9800'
            }
          },
          {
            uuid: 'group-001-0016',
            type: 'circle',
            x: 0,
            y: 0,
            w: 100,
            h: 100,
            desc: {
              bgColor: '#ffc106'
            }
          },
          {
            uuid: 'group-001-0017',
            type: 'circle',
            x: 20,
            y: 0,
            w: 100,
            h: 100,
            desc: {
              bgColor: '#cddc39'
            }
          },
          {
            uuid: 'group-001-0018',
            type: 'circle',
            x: 40,
            y: 0,
            w: 100,
            h: 100,
            desc: {
              bgColor: '#4caf50'
            }
          }
        ]
      }
    }
  ],
  modules: [],
  pages: []
};

export default data;
