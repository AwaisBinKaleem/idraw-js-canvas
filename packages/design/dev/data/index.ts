import type { DesignData } from '../../src';
import { createButton } from './components/button';
import { createCheckbox } from './components/checkbox';

const data: DesignData = {
  components: [createButton('001'), createButton('002'), createCheckbox('001'), createCheckbox('002')],
  modules: [],
  pages: []
};

export default data;
