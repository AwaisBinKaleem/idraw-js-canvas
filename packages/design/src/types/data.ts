import type { Element, ElementType, ElementSize, ElementBaseDesc } from '@idraw/types';

export type DesignComponent = Omit<ElementSize, 'angle'> & {
  uuid: string;
  type: 'component';
  name?: string;
  desc?: ElementBaseDesc & {
    children: Array<Element<ElementType> | DesignComponent>;
  };
};

export type DesignModule = Omit<ElementSize, 'angle'> & {
  uuid: string;
  type: 'module';
  name?: string;
  desc?: ElementBaseDesc & {
    children: Array<DesignComponent>;
  };
};

export type DesignPage = Omit<ElementSize, 'angle'> & {
  uuid: string;
  type: 'page';
  name?: string;
  desc: ElementBaseDesc & {
    children: Array<DesignModule>;
  };
};

export interface DesignData {
  components: DesignComponent[];
  modules: DesignModule[];
  pages: DesignPage[];
}
