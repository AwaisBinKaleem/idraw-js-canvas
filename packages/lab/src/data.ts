import type { Data } from '@idraw/types';
import { deepClone } from '@idraw/util';

const data: Data = {
  elements: [
    {
      uuid: 'xxx-0003',
      type: 'image',
      x: 100,
      y: 100,
      w: 100,
      h: 100,
      angle: 30,
      desc: {
        src: './images/lena.png'
      }
    },
    {
      uuid: 'xxxx-0001',
      x: 2,
      y: 2,
      w: 100,
      h: 100,
      type: 'circle',
      desc: {
        bgColor: '#f44336'
      }
    },
    {
      uuid: 'xxx-0002',
      type: 'rect',
      x: 50,
      y: 50,
      w: 100,
      h: 100,
      desc: {
        bgColor: '#2196f3'
      }
    },
    {
      uuid: 'xxx-0004',
      type: 'image',
      x: 250,
      y: 250,
      w: 100,
      h: 100,
      desc: {
        src: './images/github.png?t=003'
      }
    },
    {
      uuid: 'xxxx-0005',
      x: 0,
      y: 300,
      w: 100,
      h: 100,
      type: 'circle',
      desc: {
        bgColor: '#009688'
      }
    },
    {
      uuid: 'xxxx-0006',
      x: 300,
      y: 300,
      w: 100,
      h: 100,
      type: 'circle',
      desc: {
        bgColor: '#673ab7'
      }
    },
    {
      uuid: 'xxxx-0007',
      x: 300,
      y: 0,
      w: 100,
      h: 100,
      type: 'circle',
      desc: {
        bgColor: '#ffc107'
      }
    },
    {
      uuid: 'xxxx-0008',
      x: 150,
      y: 150,
      w: 100,
      h: 100,
      type: 'circle',
      desc: {
        bgColor: '#4caf50'
      }
    },
    {
      uuid: 'xxxx-0009',
      x: 0,
      y: 150,
      w: 100,
      h: 100,
      type: 'circle',
      desc: {
        bgColor: '#ff9800'
      }
    },
    {
      uuid: 'xxxx-0010',
      x: 150,
      y: 50,
      w: 100,
      h: 100,
      type: 'circle',
      desc: {
        bgColor: '#cddc39'
      }
    },
    {
      uuid: 'text-0010',
      name: 'text-002',
      x: 300,
      y: 100,
      w: 100,
      h: 60,
      type: 'text',
      desc: {
        fontSize: 16,
        text: [0, 1, 2, 3, 4].map((i) => `Hello Text ${i}`).join('\r\n'),
        // text: [0, 1, 2, 3, 4].map(i => `Hello Text ${i}`).join(''),
        fontWeight: 'bold',
        color: '#666666',
        borderRadius: 30,
        borderWidth: 2,
        borderColor: '#ff5722'
      }
    },
    {
      uuid: 'xxx-0011',
      type: 'svg',
      x: 400,
      y: 100,
      w: 100,
      h: 100,
      desc: {
        svg: `<svg viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="200" height="200"><path d="M336 421m-48 0a48 48 0 1 0 96 0 48 48 0 1 0-96 0Z" ></path><path d="M688 421m-48 0a48 48 0 1 0 96 0 48 48 0 1 0-96 0Z" ></path><path d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64z m263 711c-34.2 34.2-74 61-118.3 79.8C611 874.2 562.3 884 512 884c-50.3 0-99-9.8-144.8-29.2-44.3-18.7-84.1-45.6-118.3-79.8-34.2-34.2-61-74-79.8-118.3C149.8 611 140 562.3 140 512s9.8-99 29.2-144.8c18.7-44.3 45.6-84.1 79.8-118.3 34.2-34.2 74-61 118.3-79.8C413 149.8 461.7 140 512 140c50.3 0 99 9.8 144.8 29.2 44.3 18.7 84.1 45.6 118.3 79.8 34.2 34.2 61 74 79.8 118.3C874.2 413 884 461.7 884 512s-9.8 99-29.2 144.8c-18.7 44.3-45.6 84.1-79.8 118.2z"></path><path d="M664 533h-48.1c-4.2 0-7.8 3.2-8.1 7.4C604 589.9 562.5 629 512 629s-92.1-39.1-95.8-88.6c-0.3-4.2-3.9-7.4-8.1-7.4H360c-4.6 0-8.2 3.8-8 8.4 4.4 84.3 74.5 151.6 160 151.6s155.6-67.3 160-151.6c0.2-4.6-3.4-8.4-8-8.4z" ></path></svg>`
      }
    },
    {
      uuid: 'xxx-0012',
      x: 400,
      y: 200,
      w: 150,
      h: 100,
      type: 'html',
      angle: 0,
      desc: {
        html: `
          <style>
          .btn-box {
            margin: 10px 0;
          }
          .btn {
            line-height: 1.5715;
            position: relative;
            display: inline-block;
            font-weight: 400;
            white-space: nowrap;
            text-align: center;
            background-image: none;
            border: 1px solid transparent;
            box-shadow: 0 2px #00000004;
            cursor: pointer;
            user-select: none;
            height: 32px;
            padding: 4px 15px;
            font-size: 14px;
            border-radius: 2px;
            color: #000000d9;
            background: #fff;
            border-color: #d9d9d9;
          }
          .btn-primary {
            color: #fff;
            background: #1890ff;
            border-color: #1890ff;
            text-shadow: 0 -1px 0 rgb(0 0 0 / 12%);
            box-shadow: 0 2px #0000000b;
          }
          </style>
          <div>
            <div class="btn-box">
              <button class="btn">
                <span> Hello &nbsp; Button</span>
              </button>
            </div>
            <div class="btn-box">
              <button class="btn btn-primary">
                <span>Button Primary</span>
              </button>
            </div>
          </div>
        `
      }
    },
    {
      uuid: 'group-001',
      x: 400,
      y: 400,
      w: 100,
      h: 100,
      type: 'group',
      desc: {
        bgColor: '#1f1f1f',
        children: [
          {
            uuid: 'group-001-0014',
            type: 'circle',
            x: -40,
            y: 0,
            w: 100,
            h: 100,
            desc: {
              bgColor: '#f44336'
            }
          },
          {
            uuid: 'group-001-0015',
            type: 'circle',
            x: -20,
            y: 0,
            w: 100,
            h: 100,
            desc: {
              bgColor: '#ff9800'
            }
          },
          {
            uuid: 'group-001-0016',
            type: 'circle',
            x: 0,
            y: 0,
            w: 100,
            h: 100,
            desc: {
              bgColor: '#ffc106'
            }
          },
          {
            uuid: 'group-001-0017',
            type: 'circle',
            x: 20,
            y: 0,
            w: 100,
            h: 100,
            desc: {
              bgColor: '#cddc39'
            }
          },
          {
            uuid: 'group-001-0018',
            type: 'circle',
            x: 40,
            y: 0,
            w: 100,
            h: 100,
            desc: {
              bgColor: '#4caf50'
            }
          }
        ]
      }
    },
    {
      uuid: 'group-003',
      x: 550,
      y: 50,
      w: 173.20508075688775,
      // w: 100,
      h: 100,
      angle: 30,
      type: 'group',
      desc: {
        children: [
          {
            uuid: 'group-003-014',
            type: 'circle',
            x: -40,
            y: 0,
            w: 100,
            h: 100,
            desc: {
              bgColor: '#f44336'
            }
          },
          {
            uuid: 'group-003-0015',
            type: 'circle',
            x: -20,
            y: 0,
            w: 100,
            h: 100,
            desc: {
              bgColor: '#ff9800'
            }
          },
          {
            uuid: 'group-003-0016',
            type: 'circle',
            x: 0,
            y: 0,
            w: 100,
            h: 100,
            desc: {
              bgColor: '#ffc106'
            }
          },
          {
            uuid: 'group-003-0017',
            type: 'circle',
            x: 20,
            y: 0,
            w: 100,
            h: 100,
            desc: {
              bgColor: '#cddc39'
            }
          },
          {
            uuid: 'group-003-0018',
            type: 'circle',
            x: 40,
            y: 0,
            w: 100,
            h: 100,
            desc: {
              bgColor: '#4caf50'
            }
          }
        ]
      }
    },
    {
      uuid: 'xxxx-0017',
      type: 'image',
      x: 100,
      y: 300,
      w: 100,
      h: 100,
      angle: 30,
      desc: {
        src: './images/lena.png?v=0017'
      }
    },
    {
      uuid: 'group-004',
      x: 550,
      y: 250,
      w: 375,
      h: 400,
      type: 'group',
      desc: {
        bgColor: '#FFFFFF',
        children: [
          {
            uuid: 'groud-004-001',
            type: 'image',
            x: 200,
            y: 200,
            w: 100,
            h: 100,
            angle: 30,
            desc: {
              src: './images/lena.png'
            }
          },
          {
            uuid: 'groud-004-002',
            type: 'group',
            x: 50,
            y: 50,
            w: 200,
            h: 200,
            angle: 30,
            desc: {
              bgColor: '#f0f0f0',
              children: [
                {
                  uuid: 'group-004-002-014',
                  type: 'circle',
                  x: -40,
                  y: 0,
                  w: 100,
                  h: 100,
                  desc: {
                    bgColor: '#f44336'
                  }
                },
                {
                  uuid: 'group-004-001-0015',
                  type: 'circle',
                  x: -20,
                  y: 0,
                  w: 100,
                  h: 100,
                  desc: {
                    bgColor: '#ff9800'
                  }
                },
                {
                  uuid: 'group-004-002-0016',
                  type: 'circle',
                  x: 0,
                  y: 0,
                  w: 100,
                  h: 100,
                  desc: {
                    bgColor: '#ffc106'
                  }
                },
                {
                  uuid: 'group-004-002-0017',
                  type: 'circle',
                  x: 20,
                  y: 0,
                  w: 100,
                  h: 100,
                  desc: {
                    bgColor: '#cddc39'
                  }
                },
                {
                  uuid: 'group-004-002-0018',
                  type: 'circle',
                  x: 40,
                  y: 0,
                  w: 100,
                  h: 100,
                  desc: {
                    bgColor: '#4caf50'
                  }
                },
                {
                  uuid: 'groud-004-002-xxxx',
                  type: 'group',
                  x: 50,
                  y: 100,
                  w: 200,
                  h: 100,
                  angle: 30,
                  desc: {
                    bgColor: '#666666',
                    children: [
                      {
                        uuid: 'group-004-002-xxx-014',
                        type: 'circle',
                        x: -40,
                        y: 0,
                        w: 100,
                        h: 100,
                        desc: {
                          bgColor: '#f44336'
                        }
                      },
                      {
                        uuid: 'group-004-002-xxx-0015',
                        type: 'circle',
                        x: -20,
                        y: 0,
                        w: 100,
                        h: 100,
                        desc: {
                          bgColor: '#ff9800'
                        }
                      },
                      {
                        uuid: 'group-004-002-xxx-0016',
                        type: 'circle',
                        x: 0,
                        y: 0,
                        w: 100,
                        h: 100,
                        desc: {
                          bgColor: '#ffc106'
                        }
                      },
                      {
                        uuid: 'group-004-002-xxx-0017',
                        type: 'circle',
                        x: 20,
                        y: 0,
                        w: 100,
                        h: 100,
                        desc: {
                          bgColor: '#cddc39'
                        }
                      },
                      {
                        uuid: 'group-004-002-xxx-0018',
                        type: 'circle',
                        x: 40,
                        y: 0,
                        w: 100,
                        h: 100,
                        desc: {
                          bgColor: '#4caf50'
                        }
                      }
                    ]
                  }
                }
              ]
            }
          }
        ]
      }
    }
  ]
};

export function getData() {
  return deepClone(data);
}
