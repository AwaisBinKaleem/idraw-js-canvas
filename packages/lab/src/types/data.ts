import type { Element, ElementType, ElementSize, ElementBaseDesc } from '@idraw/types';

export type LabItemType = 'component' | 'component-item' | 'module' | 'page';

export type LabDrawDataType = 'component' | 'module' | 'page';

export type LabComponentItem = ElementSize & {
  uuid: string;
  type: 'component-item';
  name: string;
  detail?: ElementBaseDesc & {
    children: Array<Element<ElementType> | LabComponentItem>;
  };
};

export type LabComponent = ElementSize & {
  uuid: string;
  type: 'component';
  name: string;
  detail?: ElementBaseDesc & {
    default: LabComponentItem;
    variants: LabComponentItem[];
  };
};

export type LabModule = ElementSize & {
  uuid: string;
  type: 'module';
  name: string;
  detail?: ElementBaseDesc & {
    children: Array<LabComponent>;
  };
};

export type LabPage = ElementSize & {
  uuid: string;
  type: 'page';
  name: string;
  detail: ElementBaseDesc & {
    children: Array<LabModule>;
  };
};

export interface LabData {
  components: LabComponent[];
  modules: LabModule[];
  pages: LabPage[];
}
