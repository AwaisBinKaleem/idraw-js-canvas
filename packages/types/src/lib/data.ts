import type { Element, ElementType, ElementAssets, ElementSize, ElementGroupDetail } from './element';

export type DataLayout = Pick<ElementSize, 'w' | 'h'> & {
  detail: Omit<ElementGroupDetail, 'children'>;
};
export interface Data<E extends Record<string, any> = Record<string, any>> {
  elements: Element<ElementType, E>[];
  assets?: ElementAssets;
  layout?: DataLayout;
}

export type Matrix = [
  number,
  number,
  number,
  number,
  number,
  number,
  number,
  number,
  number,
  number,
  number,
  number,
  number,
  number,
  number,
  number,
  number,
  number,
  number,
  number
];

export type ColorMatrix = Matrix;
