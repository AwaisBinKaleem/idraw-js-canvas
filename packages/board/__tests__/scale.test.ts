import Board from '../src';
import { getData } from './data';

describe('@idraw/board', () => {

  test('scale', async () => {  
    document.body.innerHTML = `
      <div id="mount"></div>
    `;
    const opts = {
      width: 600,
      height: 400,
      contextWidth: 600,
      contextHeight: 400,
      devicePixelRatio: 4
    }
    const mount = document.querySelector('#mount') as HTMLDivElement;
    const board = new Board(mount, opts);
  
    const ctx = board.getContext();
    const data = getData();
  
    board.clear();
    ctx.clearRect(0, 0, opts.width, opts.height);
    ctx.setFillStyle('#ffffff');
    ctx.fillRect(0, 0, opts.width, opts.height);
    data.elements.forEach(ele => {
      ctx.setFillStyle(ele.desc.color);
      ctx.fillRect(ele.x, ele.y, ele.w, ele.h);
    });
  
    const result = board.scale(0.5);
    expect(result).toStrictEqual({"position":{"top":150,"bottom":100,"left":100,"right":150},"size":{"x":75,"y":50,"w":300,"h":200}})
    board.draw();
  
    const originCtx = board.getOriginContext();
    // @ts-ignore;
    const originCalls = originCtx.__getDrawCalls();
    expect(originCalls).toMatchSnapshot();
  
    const displayCtx = board.getDisplayContext();
    // @ts-ignore;
    const displayCalls = displayCtx.__getDrawCalls();
    expect(displayCalls).toMatchSnapshot();
  
  });
})
