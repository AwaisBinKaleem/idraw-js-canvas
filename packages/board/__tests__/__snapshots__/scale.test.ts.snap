// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@idraw/board scale 1`] = `
Array [
  Object {
    "props": Object {
      "height": 1600,
      "width": 2400,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "height": 1600,
      "width": 2400,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "dHeight": 1600,
      "dWidth": 2400,
      "dx": 600,
      "dy": 400,
      "img": <canvas
        height="1600"
        width="2400"
      />,
      "sHeight": 1600,
      "sWidth": 2400,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;

exports[`@idraw/board scale 2`] = `
Array [
  Object {
    "props": Object {
      "height": 1600,
      "width": 2400,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "height": 1600,
      "width": 2400,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "dHeight": 1600,
      "dWidth": 2400,
      "dx": 600,
      "dy": 400,
      "img": <canvas
        height="1600"
        width="2400"
      />,
      "sHeight": 1600,
      "sWidth": 2400,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;
